x-pack/auditbeat/processors/add_session_metadata/pkg/processdb/entry_leader_test.go

```diff
@@ -0,0 +1,1247 @@
+	require.Nil(t, process.Parent.Start)
+}
+
+func requireSessionLeader(t *testing.T, db DB, pid uint32, sid uint32) {
+	process, err := db.GetProcess(pid)
+	require.Nil(t, err)
+	require.Equal(t, sid, process.SessionLeader.PID)
+	require.NotNil(t, process.SessionLeader.SameAsProcess)
+	require.Equal(t, pid == sid, *process.SessionLeader.SameAsProcess)
+}
+
+func requireSessionLeaderUnset(t *testing.T, process types.Process) {
+	require.Equal(t, "", process.SessionLeader.EntityID)
+	require.Equal(t, uint32(0), process.SessionLeader.PID)
+	require.Nil(t, process.SessionLeader.Start)
+}
+
+func requireGroupLeader(t *testing.T, db DB, pid uint32, pgid uint32) {
+	process, err := db.GetProcess(pid)
+	require.Nil(t, err)
+	require.Equal(t, pgid, process.GroupLeader.PID)
+	require.NotNil(t, process.GroupLeader.SameAsProcess)
+	require.Equal(t, pid == pgid, *process.GroupLeader.SameAsProcess)
+}
+
+func requireEntryLeader(t *testing.T, db DB, pid uint32, entryPid uint32, expectedEntryType EntryType) {
+	process, err := db.GetProcess(pid)
+	require.Nil(t, err)
+	require.Equal(t, entryPid, process.EntryLeader.PID)
+	require.NotNil(t, process.EntryLeader.SameAsProcess)
+	require.Equal(t, pid == entryPid, *process.EntryLeader.SameAsProcess)
+
+	entryType, err := db.GetEntryType(entryPid)
+	require.Nil(t, err)
+	require.Equal(t, expectedEntryType, entryType)
+}
+
+func requireEntryLeaderUnset(t *testing.T, process types.Process) {
+	require.Equal(t, "", process.EntryLeader.EntityID)
+	require.Equal(t, uint32(0), process.EntryLeader.PID)
+	require.Nil(t, process.EntryLeader.Start)
+}
+
+// tries to construct fork event from what's in the db
+func insertForkAndExec(t *testing.T, db DB, exec types.ProcessExecEvent) {
+	var fork types.ProcessForkEvent
+	fork.ChildPids = exec.Pids
+	parent, err := db.GetProcess(exec.Pids.Ppid)
+	if err != nil {
+		fork.ParentPids = exec.Pids
+		fork.ParentPids.Tgid = exec.Pids.Ppid
+		fork.ParentPids.Ppid = 0
+		fork.ParentPids.Pgid = 0
+
+		fork.ChildPids.Pgid = exec.Pids.Ppid
+
+		// if the exec makes itself a session and the parent is no where to be
+		// found we'll make the parent its own session
+		if exec.Pids.Tgid == exec.Pids.Sid {
+			fork.ParentPids.Sid = exec.Pids.Ppid
+		}
+	} else {
+		fork.ParentPids.Tgid = parent.PID
+		fork.ParentPids.Ppid = parent.Parent.PID
+		fork.ParentPids.Sid = parent.SessionLeader.PID
+
+		// keep group leader the same for now
+		fork.ParentPids.Pgid = exec.Pids.Pgid
+	}
+
+	if fork.ParentPids.Tgid != 0 {
+		err = db.InsertFork(fork)
+		require.Nil(t, err)
+	}
+
+	err = db.InsertExec(exec)
+	require.Nil(t, err)
+}
+
+var systemdPath = "/sbin/systemd"
+
+func populateProcfsWithInit(reader *procfs.MockReader) {
+	reader.AddEntry(1, procfs.ProcessInfo{
+		Pids: types.PidInfo{
+			Tid: 1,
+			Tgid: 1,
+			Pgid: 0,
+			Sid: 1,
+		},
+		Filename: systemdPath,
+	})
+}
+
+func testSingleProcessSessionLeaderEntryTypeTerminal(createDB createDBFn) testFn {
```

Comment:
Can these tests be done with tables? It looks to me like they can. rsc gave a good talk recently on Go testing [here](https://research.swtch.com/testing).